features/timeseries/src/main/java/org/opennms/netmgt/timeseries/integration/TimeseriesWriter.java

```diff
@@ -0,0 +1,298 @@
+                                    .distinct()
+                                    .collect(Collectors.joining(", "));
+                        }
+                    });
+            droppedSamples.mark(samples.size());
+            return;
+        }
+        // Increase our entry counter
+        numEntriesOnRingBuffer.incrementAndGet();
+    }
+
+    @Override
+    public void onEvent(SampleBatchEvent event) throws Exception {
+        // We'd expect the logs from this thread to be in collectd.log
+        Logging.putPrefix("collectd");
+
+        // Decrement our entry counter
+        numEntriesOnRingBuffer.decrementAndGet();
+
+        if (event.isIndexOnly()) {
+            storeMetadata(event);
+        } else {
+            storeTimeseriesData(event);
+            storeMetadata(event);
+        }
+    }
+
+    private void storeTimeseriesData(SampleBatchEvent event) throws StorageException {
+        List<org.opennms.integration.api.v1.timeseries.Sample> samples
+                = event.getSamples().stream().map(this::toApiSample).collect(Collectors.toList());
+        this.storage.get().store(samples);
+    }
+
+    private void storeMetadata(SampleBatchEvent event) throws SQLException {
+        // dedouble attributes
+        Set<MetaData> metaData = new HashSet<>();
+        for(Sample sample : event.getSamples()) {
+
+            // attributes of sample
+            if(sample.getResource().getAttributes().isPresent()) {
+                sample.getResource().getAttributes().get().forEach((key, value) -> metaData.add(new MetaData(sample.getResource().getId(), key, value)));
+            }
+        }
+        this.timeSeriesMetaDataDao.store(metaData);
+    }
+
+    private org.opennms.integration.api.v1.timeseries.Sample toApiSample(final Sample sample) {
+
+        ImmutableMetric.MetricBuilder builder = ImmutableMetric.builder()
+                .intrinsicTag(CommonTagNames.resourceId, sample.getResource().getId())
+                .intrinsicTag(CommonTagNames.name, sample.getName())
+                .metaTag(typeToTag(sample.getType()));
+
+        if(sample.getResource().getAttributes().isPresent()) {
+            sample.getResource().getAttributes().get().forEach(builder::metaTag);
+        }
+
+        final ImmutableMetric metric = builder.build();
+        final Instant time = Instant.ofEpochMilli(sample.getTimestamp().asMillis());
+        final Double value = sample.getValue().doubleValue();
+
+        return ImmutableSample.builder().metric(metric).time(time).value(value).build();
+    }
+
+    private Tag typeToTag (MetricType type) {
+        ImmutableMetric.Mtype mtype;
+        if(type == MetricType.GAUGE){
+            mtype = ImmutableMetric.Mtype.gauge;
+        } else if(type == MetricType.COUNTER) {
+            mtype = ImmutableMetric.Mtype.count;
+        } else {
+            throw new IllegalArgumentException("Implement me"); // no other types exist
```

Comment:
Exception should include the type in the message.

Reply:
:+1: done